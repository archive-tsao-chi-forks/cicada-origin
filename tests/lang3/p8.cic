List : (Type) -> Type
List = (T) => { Null(T) | Cons(T) }

Null : (Type) -> Type
Null = (T) => Object

Cons : (Type) -> Type
Cons = (T) => {
  head: T
  tail: List(T)
}

@show List

_ : Null(String)
_ = {}

@show _

// _ : List(String)
// _ = {}

// @show _

// _ : Dup
// _ = {
//   T = String
//   x = "abc"
//   y = "123"
// }

// @show _

// _ : Dup[String]
// _ = {
//   T = String
//   x = "abc"
//   y = "123"
// }

// @show _

// _ : Dup[String]["abc"]
// _ = {
//   T = String
//   x = "abc"
//   y = "123"
// }

// @show _

// _ : Dup[String]["abc"]["123"]
// _ = {
//   T = String
//   x = "abc"
//   y = "123"
// }

// @show _
