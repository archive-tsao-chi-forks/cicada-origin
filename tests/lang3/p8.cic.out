(_: Type) -> Type -- (_) => { Object | {
  head : _
  tail : { Object | {
    head : _
    tail : List(_)
  } }
} }
Object -- {
  
}
{ Object | {
  head : String
  tail : { Object | {
    head : String
    tail : List(String)
  } }
} } -- {
  
}
